gh-109595: Add -Xcpu_count=<n> cmdline for container users

corona10 opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-109595 -->
* Issue: gh-109595
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
:books: Documentation preview :books:: https://cpython-previews--109667.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

corona10 committed:
gh-109595: Add -Xcpu_count=<n> cmdline for container users

corona10 committed:
Check style

corona10 committed:
Add help

vstinner commented on Doc/library/os.rst:
```diff
@@ -5164,6 +5164,10 @@ Miscellaneous System Information
 
+   .. versionchanged:: 3.13
+      if ``-X cpu_count=<n>`` is given, :func:`cpu_count` will return the overrided
+      value.
```
Can you integrate this text in the doc above, and here just mention that the function takes -X cpu_count in account?

```suggestion
      if :samp:`-X cpu_count={n}` is given, :func:`cpu_count` will return the overrided
      value *n*.
```

vstinner commented on Doc/using/cmdline.rst:
```diff
@@ -546,6 +546,10 @@ Miscellaneous options
      will do nothing if is not supported on the current system. The default value
      is "off". See also :envvar:`PYTHONPERFSUPPORT` and :ref:`perf_profiling`.
+   * ``-X cpu_count=<n>`` will override the number of cpu count from system.
+     If this option is provided, :func:`os.cpu_count` will return the overrided value.
```
> will override the number of cpu count from system

Honestly, that's vague. Just describe what it does:

```suggestion
   * ``-X cpu_count=<n>`` overrides :func:`os.cpu_count`.
```

vstinner commented on Doc/using/cmdline.rst:
```diff
@@ -546,6 +546,10 @@ Miscellaneous options
      is "off". See also :envvar:`PYTHONPERFSUPPORT` and :ref:`perf_profiling`.
+   * ``-X cpu_count=<n>`` will override the number of cpu count from system.
+     If this option is provided, :func:`os.cpu_count` will return the overrided value.
+     This option will be useful for users who need to limit cpu resource of the container system.
```
```suggestion
     This option will be useful for users who need to limit CPU resources of a container system.
```

Please document that *n* must be at least 1.

vstinner commented on Doc/using/cmdline.rst:
```diff
@@ -593,6 +597,9 @@ Miscellaneous options
       The ``-X perf`` option.
 
+   .. versionadded:: 3.13
+      The ``-X cpu_count`` option
```
```suggestion
      The ``-X cpu_count`` option.
```

vstinner commented on Include/cpython/initconfig.h:
```diff
@@ -179,6 +179,7 @@ typedef struct PyConfig {
     int use_frozen_modules;
     int safe_path;
     int int_max_str_digits;
+    int cpu_count;
```
Maybe we should start adding version to this complex structure:

```suggestion

    // Python 3.13
    int cpu_count;
```

erlend-aasland replied:
Would it not be better to document the feature instead of the version? Using simple Git tricks, it is very easy to find out which Python version a field was added in; I don't see the value in duplicating Git log in the source code :)

vstinner replied:
All PyConfig must be documented at: https://docs.python.org/dev/c-api/init_config.html

See my notes on adding a new member to PyConfig: https://pythondev.readthedocs.io/pyconfig.html

erlend-aasland replied:
> All PyConfig must be documented at: https://docs.python.org/dev/c-api/init_config.html

Yes, so IMO:

- remove the superfluous `// Python 3.13` comment from initconfig.h
- add an entry accompanied with a `.. versionadded` to [Doc/c-api/init_config.rst](https://github.com/python/cpython/blob/main/Doc/c-api/init_config.rst)

vstinner replied:
Sure, in the doc, versionadded must be used, obviously.

> remove the superfluous // Python 3.13 comment from initconfig.h

I think that it would nice to *also* document the version in PyConfig structure, in addition to the init_config.rst doc. But I will not insist :-)

erlend-aasland replied:
Will a `// Python 3.13` comment make sense on its own? IMO: no. How would a random contributor interpret it? :) How will you or I interpret it in five years time? Isn't it better to leave the version control to the version control system? :)

erlend-aasland replied:
... so you can count me -1 on sprinkling the source code with `// Python x.y` comments :)

vstinner replied:
Ok, let's remove the "// Python 3.13" comment.

vstinner commented on Lib/test/test_cmd_line.py:
```diff
@@ -901,6 +901,14 @@ def res2int(res):
+            return tuple(int(i) for i in out.split())
+        code = "import os; print(os.cpu_count())"
+        res = assert_python_ok('-X', 'cpu_count=4321', '-c', code)
+        self.assertEqual(res2int(res), (4321,))
```
Why not just testing that you get ``b"4321\n"``? Is it useful to parse the string?

I hate assert_python_ok() API which doesn't let using text=True parameter.

vstinner commented on Misc/NEWS.d/next/Core and Builtins/2023-09-22-01-44-53.gh-issue-109595.fVINgD.rst:
```diff
@@ -0,0 +1,2 @@
+Add ``-Xcpu_count=<n>`` cmdline option for limiting system cpu resources from
```
```suggestion
Add :samp:`-Xcpu_count={n}` command line option for limiting system CPU resources from
```

Hum, maybe just say that it overrides os.cpu_count()?

vstinner commented on Modules/posixmodule.c:
```diff
@@ -14349,6 +14349,11 @@ static PyObject *
 /*[clinic end generated code: output=5fc29463c3936a9c input=e7c8f4ba6dbbadd3]*/
 {
+    PyThreadState *tstate = _PyThreadState_GET();
+    const PyConfig *config = _PyInterpreterState_GetConfig(tstate->interp);
```
```suggestion
    const PyConfig *config = _Py_GetConfig();
```

vstinner commented on Python/initconfig.c:
```diff
@@ -789,6 +793,7 @@ PyConfig_InitIsolatedConfig(PyConfig *config)
     config->int_max_str_digits = _PY_LONG_DEFAULT_MAX_STR_DIGITS;
     config->safe_path = 1;
     config->pathconfig_warnings = 0;
+    config->cpu_count = -1;
```
it's already set by config_init_defaults(), no ned to repeat the default here.

vstinner commented on Python/initconfig.c:
```diff
@@ -1678,6 +1684,28 @@ config_read_env_vars(PyConfig *config)
+    }
+    return _PyStatus_OK();
+error:
+    return _PyStatus_ERR("-X cpu_count=<n> invalid cpu_count format");
```
```suggestion
    return _PyStatus_ERR("-X cpu_count=n option: n is missing or an invalid number");
```

vstinner commented on Doc/using/cmdline.rst:
```diff
@@ -546,6 +546,10 @@ Miscellaneous options
      report Python calls. This option is only available on some platforms and
      will do nothing if is not supported on the current system. The default value
      is "off". See also :envvar:`PYTHONPERFSUPPORT` and :ref:`perf_profiling`.
+   * ``-X cpu_count=<n>`` will override the number of cpu count from system.
```
```suggestion
   * :samp:`-X cpu_count={n}` will override the number of CPU count from system.
```

vstinner commented on Python/initconfig.c:
```diff
@@ -130,7 +130,10 @@ The following implementation-specific options are available:\n\
-    The default is sys.int_info.default_max_str_digits.  0 disables."
+    The default is sys.int_info.default_max_str_digits.  0 disables.\n\
+\n\
+-X cpu_count=number: limit cpu count of os.cpu_count.\n\
```
If you prefer to write "number" instead of "n", please update the whole PR for that.

```suggestion
-X cpu_count=n: limit CPU count of os.cpu_count().\n\
```

Why is it a "limit"? Your unit test shows that it can be BIGGER than the real number of logical CPUs of the system!

Maybe just say "override"?

vstinner commented on Python/initconfig.c:
```diff
@@ -130,7 +130,10 @@ The following implementation-specific options are available:\n\
+    The default is sys.int_info.default_max_str_digits.  0 disables.\n\
+\n\
+-X cpu_count=number: limit cpu count of os.cpu_count.\n\
+    This helps for users who need to limit cpu resource of the container system."
```
```suggestion
    This helps for users who need to limit CPU resources of a container system."
```

vstinner commented on Python/initconfig.c:
```diff
@@ -959,6 +964,7 @@ _PyConfig_Copy(PyConfig *config, const PyConfig *config2)
     COPY_WSTRLIST(orig_argv);
     COPY_ATTR(_is_python_build);
     COPY_ATTR(int_max_str_digits);
+    COPY_ATTR(cpu_count);
```
Search for "int_max_str_digits", you should also modify:

* config_check_consistency()
* _PyConfig_AsDict()
* _PyConfig_FromDict()

You should update test_embed, and maybe set cpu_count in test_init_from_config() of Programs/_testembed.c.

See also my notes: https://pythondev.readthedocs.io/pyconfig.html

vstinner commented on Python/initconfig.c:
```diff
@@ -1678,6 +1684,28 @@ config_read_env_vars(PyConfig *config)
+        config->cpu_count = cpu_count;
+    }
+    return _PyStatus_OK();
+error:
```
```suggestion
    return _PyStatus_OK();
    
error:
```

corona10 committed:
Apply suggestions from code review

Co-authored-by: Victor Stinner <vstinner@python.org>

corona10 committed:
Apply suggestions from code review

Co-authored-by: Victor Stinner <vstinner@python.org>

corona10 committed:
Check style

corona10 committed:
Address code review

corona10 committed:
Address code review

corona10 committed:
Address code review

corona10 committed:
Address code review

corona10 committed:
Fix test

corona10 committed:
Update NEWS.d

vstinner commented on Programs/_testembed.c:
```diff
@@ -715,6 +715,7 @@ static int test_init_from_config(void)
 
     putenv("PYTHONINTMAXSTRDIGITS=6666");
     config.int_max_str_digits = 31337;
+    config.cpu_count = -1;
```
Can you please set a more interesting value like 1234?

corona10 committed:
Address code review

corona10 committed:
Update

vstinner commented on Doc/using/cmdline.rst:
```diff
@@ -546,6 +546,11 @@ Miscellaneous options
      is "off". See also :envvar:`PYTHONPERFSUPPORT` and :ref:`perf_profiling`.
+   * :samp:`-X cpu_count={n}` will override the number of CPU count from system.
+     If this option is provided, :func:`os.cpu_count` will return the overrided value.
+     And *n* must be greater equal then 1.
```
```suggestion
   * :samp:`-X cpu_count={n}` overrides the number of CPU count from system:
     :func:`os.cpu_count` returns *n*. The value *n* must be greater than
     or equal to 1.
```

corona10 committed:
Update

vstinner commented on Doc/using/cmdline.rst:
```diff
@@ -546,6 +546,11 @@ Miscellaneous options
+   * :samp:`-X cpu_count={n}` will override the number of CPU count from system.
+     If this option is provided, :func:`os.cpu_count` will return the overrided value.
+     And *n* must be greater equal then 1.
+     This option will be useful for users who need to limit CPU resources of a container system.
```
```suggestion
     This option is useful for users who need to limit CPU resources of a container system.
```

corona10 committed:
nit

vstinner commented on Misc/NEWS.d/next/Core and Builtins/2023-09-22-01-44-53.gh-issue-109595.fVINgD.rst:
```diff
@@ -0,0 +1,2 @@
+Add :samp:`-Xcpu_count={n}` command line option for overriding system CPU resources from
+Python program. Patch by Donghee Na.
```
```suggestion
Add :samp:`-Xcpu_count={n}` command line option to override CPU count of :func:`os.cpu_count`.
This option is useful for users who need to limit CPU resources of a container system.
Patch by Donghee Na.
```

corona10 committed:
Update

vstinner commented on Python/initconfig.c:
```diff
@@ -633,6 +636,8 @@ config_check_consistency(const PyConfig *config)
     assert(config->safe_path >= 0);
     assert(config->int_max_str_digits >= 0);
+    // cpu_count can be -1 if the user doesn't override it.
+    assert(config->cpu_count != 0);
```
Oh, that's interesting that the value cannot be zero :-)

vstinner commented on Python/initconfig.c:
```diff
@@ -1678,6 +1687,29 @@ config_read_env_vars(PyConfig *config)
+static PyStatus
+config_init_cpu_count(PyConfig *config)
+{
+    int cpu_count = -1;
```
You can move the variable declaration inside the "if (sep)" block. Does it *have* to be initialized?

If you are afraid of undefined behavior, maybe config_wstr_to_int() should set ``*result`` to 0 on error.

corona10 replied:
Now I added `PYTHONCPUCOUNT` envvar, so the current structure will be needed.

vstinner replied:
If you want, you *can* move the variable declaration inside each "if (env)" block and duplicate the variable, to better show its scope. But well, that's just a personal preference. Feel free to ignore my coding style remark ;-)

vstinner commented on Python/initconfig.c:
```diff
@@ -1678,6 +1687,29 @@ config_read_env_vars(PyConfig *config)
+    return _PyStatus_OK();
+
+error:
+    return _PyStatus_ERR("-X cpu_count=n option: n is missing or an invalid number");
```
```suggestion
    return _PyStatus_ERR("-X cpu_count=n option: n is missing or an invalid number, ");
                         "n must be greater than 0");    
```

corona10 committed:
Add PYTHONCPUCOUNT

vstinner reviewed (commented):
Can you please document the change in Doc/whatsnew/3.13.rst?

corona10 committed:
Add PYTHONCPUCOUNT

corona10 committed:
Apply suggestions from code review

Co-authored-by: Victor Stinner <vstinner@python.org>

corona10 committed:
Update Python/initconfig.c

Co-authored-by: Victor Stinner <vstinner@python.org>

corona10 committed:
Fix

corona10 committed:
Update whatsnew

corona10 committed:
nit

corona10 committed:
Check style

corona10 committed:
Fix docs

corona10 committed:
Update doc

corona10 committed:
nit

corona10 committed:
Update

corona10 committed:
Update

corona10 committed:
Update

vstinner commented on Doc/using/cmdline.rst:
```diff
@@ -546,6 +546,11 @@ Miscellaneous options
      will do nothing if is not supported on the current system. The default value
      is "off". See also :envvar:`PYTHONPERFSUPPORT` and :ref:`perf_profiling`.
+   * :samp:`-X cpu_count={n}` overrides :func:`os.cpu_count`.
+     And *n* must be greater than 0.
```
```suggestion
     *n* must be greater than or equal to 1.
```

I prefer >= 1, it's more human friendly :-)

vstinner commented on Doc/whatsnew/3.13.rst:
```diff
@@ -163,6 +163,14 @@ opcode
+--
+
+* :func:`os.cpu_count` can be overrided  through the new
+  environment variable :envvar:`PYTHONCPUCOUNT`, the new command-line option
```
```suggestion
  environment variable :envvar:`PYTHONCPUCOUNT` and the new command-line option
```

vstinner commented on Lib/test/test_cmd_line.py:
```diff
@@ -901,6 +901,11 @@ def res2int(res):
+    def test_cpu_count(self):
+        code = "import os; print(os.cpu_count())"
+        res = assert_python_ok('-X', 'cpu_count=4321', '-c', code)
+        self.assertEqual(res.out.strip().decode("utf-8"), '4321')
```
That's a lot of CPUs, it's exciting! Until *recently*, Linux was limited to 4096 CPUs! https://unix.stackexchange.com/questions/4507/how-many-cores-can-linux-kernel-handle

vstinner commented on Modules/posixmodule.c:
```diff
@@ -14349,6 +14349,10 @@ static PyObject *
+    if (config->cpu_count > 0) {
+        return PyLong_FromLong(config->cpu_count);
+    }
     int ncpu = 0;
```
```suggestion
    }
    
    int ncpu = 0;
```

vstinner commented on Python/initconfig.c:
```diff
@@ -1678,6 +1687,39 @@ config_read_env_vars(PyConfig *config)
+            config->cpu_count = cpu_count;
+        }
+    }
+    const wchar_t *xoption = config_get_xoption(config, L"cpu_count");
```
```suggestion
    }
    
    const wchar_t *xoption = config_get_xoption(config, L"cpu_count");
```

corona10 committed:
Address code review

corona10 committed:
Address code review

corona10 committed:
Address Victor's suggestion

corona10 committed:
nit

corona10 committed:
Address Erlend's review

corona10 committed:
fix

corona10 committed:
Merge remote-tracking branch 'upstream/main' into gh-109595

corona10 committed:
fix

corona10 committed:
Add space

corona10 committed:
Update os.py

corona10 committed:
nit

corona10 committed:
Add test code

corona10 committed:
Update docs

vstinner commented on Doc/c-api/init_config.rst:
```diff
@@ -878,6 +878,18 @@ PyConfig
 
+   .. c:member:: int cpu_count
+
+      If the value of <cpu_count> is not ``-1`` then it will override
```
This syntax emits a warning in Doc. Use ``*cpu_count*`` or ``:c:member:`~PyConfig.cpu_count` ``.

vstinner commented on Doc/c-api/init_config.rst:
```diff
@@ -878,6 +878,18 @@ PyConfig
+   .. c:member:: int cpu_count
+
+      If the value of <cpu_count> is not ``-1`` then it will override
+      the return value of :func:`cpu_count` and :func:`process_cpu_count` into <cpu_count>.
```
```suggestion
      the return value of :func:`os.cpu_count` and :func:`os.process_cpu_count` into *cpu_count*.
```

vstinner commented on Doc/c-api/init_config.rst:
```diff
@@ -878,6 +878,18 @@ PyConfig
+      Configured by the :samp:`-X cpu_count={n}` command line
+      flag or the :envvar:`PYTHONCPUCOUNT` environment variable.
+
+      Default: ``-1``, it will follow the original behavior of :func:`cpu_count` and :func:`process_cpu_count.
```
```suggestion
      Default: ``-1``.
```

vstinner commented on Doc/c-api/init_config.rst:
```diff
@@ -878,6 +878,18 @@ PyConfig
+      the return value of :func:`cpu_count` and :func:`process_cpu_count` into <cpu_count>.
+
+      Configured by the :samp:`-X cpu_count={n}` command line
+      flag or the :envvar:`PYTHONCPUCOUNT` environment variable.
```
Please document ``default`` value.

vstinner commented on Doc/using/cmdline.rst:
```diff
@@ -546,6 +546,13 @@ Miscellaneous options
+     *n* must be greater than or equal to 1.
+     This option is useful for users who need to limit CPU resources of a container system.
+     See also :envvar:`PYTHONCPUCOUNT`.
+     If *n* is "default", it follows default :func:`os.cpu_count` even if :envvar:`PYTHONCPUCOUNT`
```
```suggestion
     If *n* is ``default``, it follows default :func:`os.cpu_count` even if :envvar:`PYTHONCPUCOUNT`
```

vstinner commented on Doc/using/cmdline.rst:
```diff
@@ -1063,6 +1073,15 @@ conflict.
+.. envvar:: PYTHONCPUCOUNT
+
+   If this variable is set to a positive integer, it overrides
+   :func:`os.cpu_count` and and :func:`os.process_cpu_count`.
```
```suggestion
   :func:`os.cpu_count` and and :func:`os.process_cpu_count` result.
```

vstinner commented on Doc/using/cmdline.rst:
```diff
@@ -1063,6 +1073,15 @@ conflict.
+   If this variable is set to a positive integer, it overrides
+   :func:`os.cpu_count` and and :func:`os.process_cpu_count`.
+
+   See also the :samp:`-X cpu_count={n}` command-line option.
```
```suggestion
   See also the :opt:`-X cpu_count <-X>` command-line option.
```

vstinner commented on Doc/whatsnew/3.13.rst:
```diff
@@ -166,6 +166,11 @@ opcode
+* :func:`os.cpu_count` and :func:`os.process_cpu_count` can be overrided through
+  the new environment variable :envvar:`PYTHONCPUCOUNT` or the new command-line option
+  :option:`-X cpu_count <-X>`. This option is useful for users who need to limit
+  CPU resources of a container system.(Contributed by Donghee Na in :gh:`109595`)
```
Document it *after* the item below which documents the addition of process_cpu_count().

vstinner replied:
```suggestion
  CPU resources of a container system.
  (Contributed by Donghee Na in :gh:`109595`)
```

vstinner commented on Lib/test/test_cmd_line.py:
```diff
@@ -902,6 +902,26 @@ def res2int(res):
+    def test_process_cpu_count(self):
+        code = "import os; print(os.process_cpu_count())"
+        res = assert_python_ok('-X', 'cpu_count=4321', '-c', code)
+        self.assertEqual(res.out.strip().decode("utf-8"), '4321')
```
Honestly, you can merged this test with test_cpu_count(): test the two functions at once:

```
code = "import os; print(os.cpu_count(), os.process_cpu_count())"
```

vstinner commented on Lib/test/test_cmd_line.py:
```diff
@@ -902,6 +902,26 @@ def res2int(res):
+    def test_cpu_count_default(self):
+        code = "import os; print(os.cpu_count())"
+        res = assert_python_ok('-X', 'cpu_count=default', '-c', code)
+        self.assertEqual(int(res.out.strip().decode("utf-8")), os.cpu_count())
```
You should set the env var to a value.

corona10 committed:
fix

vstinner commented on Lib/os.py:
```diff
@@ -1138,7 +1138,17 @@ def add_dll_directory(path):
+        return False
+    if xopt == 'default':
+        return False
+    return True
```
I would prefer to expose PyConfig.cpu_count value using a private function, rather than having two functions parsing env var / cmdline option. Otherwise, the two implementations may become inconsistent later. Also, PyConfig can be modified with the PyConfig API, not only using env var / cmdline.

I'm even working on configuration file-like API: https://github.com/python/cpython/pull/110145

corona10 committed:
Address code reivew

corona10 committed:
Address code review

corona10 committed:
Add test

corona10 committed:
fix

corona10 committed:
fix

corona10 committed:
fix

corona10 committed:
fix

vstinner commented on Doc/c-api/init_config.rst:
```diff
@@ -880,13 +880,13 @@ PyConfig
-      If the value of <cpu_count> is not ``-1`` then it will override
-      the return value of :func:`cpu_count` and :func:`process_cpu_count` into <cpu_count>.
+      If the value of :c:member:`~PyConfig.cpu_count` is not ``-1`` then it will override
+      the return value of :func:`os.cpu_count` and :func:`os.process_cpu_count` into <cpu_count>.
```
```suggestion
      the return value of :func:`os.cpu_count` and :func:`os.process_cpu_count` functions
      into *cpu_count*.
```

corona10 committed:
Update

vstinner commented on Doc/using/cmdline.rst:
```diff
@@ -550,7 +550,7 @@ Miscellaneous options
      This option is useful for users who need to limit CPU resources of a container system.
      See also :envvar:`PYTHONCPUCOUNT`.
-     If *n* is "default", it follows default :func:`os.cpu_count` even if :envvar:`PYTHONCPUCOUNT`
+     If *n* is ``default``, it follows default :func:`os.cpu_count` even if :envvar:`PYTHONCPUCOUNT`
```
I suggest:

```suggestion
     If *n* is ``default``, :func:`os.cpu_count` and  :func:`os.process_cpu_count` are not overridden.
```

vstinner commented on Doc/using/cmdline.rst:
```diff
@@ -1076,9 +1076,9 @@ conflict.
 
    If this variable is set to a positive integer, it overrides
-   :func:`os.cpu_count` and and :func:`os.process_cpu_count`.
+   :func:`os.cpu_count` and and :func:`os.process_cpu_count` result.
```
In init_config, you used "result". I would prefer consistency, either "result" in both docs, or "return value" in both docs.

```suggestion
   :func:`os.cpu_count` and and :func:`os.process_cpu_count` return value.
```

vstinner commented on Doc/whatsnew/3.13.rst:
```diff
@@ -166,15 +166,16 @@ opcode
+* :func:`os.cpu_count` and :func:`os.process_cpu_count` can be overrided through
+  the new environment variable :envvar:`PYTHONCPUCOUNT` or the new command-line option
+  :option:`-X cpu_count <-X>`. This option is useful for users who need to limit
+  CPU resources of a container system.
```
```suggestion
  CPU resources of a container system without having to modify the container (application code).
```

vstinner commented on Lib/test/test_cmd_line.py:
```diff
@@ -878,49 +878,39 @@ def test_int_max_str_digits(self):
-            out = res.out.strip().decode("utf-8")
-            return tuple(int(i) for i in out.split())
-
         res = assert_python_ok('-c', code)
```
nitpick: you can use res2int = self.res2int, to keep the tests unchanged.

corona10 committed:
Address code review

corona10 committed:
Update NEWS.d

corona10 committed:
Update

corona10 committed:
Update

vstinner commented on Python/initconfig.c:
```diff
@@ -228,7 +229,10 @@ The following implementation-specific options are available:\n\
-    The default is sys.int_info.default_max_str_digits.  0 disables."
+    The default is sys.int_info.default_max_str_digits.  0 disables.\n\
+\n\
+-X cpu_count=n: override CPU count of os.cpu_count().\n\
```
and process_cpu_count(). You may document "default" as well.

vstinner commented on Python/initconfig.c:
```diff
@@ -1615,6 +1622,44 @@ config_read_env_vars(PyConfig *config)
+            cpu_count = -1;
+        }
+        if (cpu_count >= 1) {
+            config->cpu_count = cpu_count;
```
What about raising an error, rather than silently ignoring invalid value?

For consistency, you should accept default.

corona10 committed:
Address code review

corona10 committed:
Rename to PYTHON_CPU_COUNT

vstinner reviewed (approved):
LGTM. I prefer ``PYTHON_CPU_COUNT`` env name, even if sadly it breaks an old bad habbit of "no underscores".

@gpshead: Would you mind to have a look at this change? Are you ok with the overall approach?

corona10 committed:
Hidden overrided cpu count

corona10 committed:
Use overridden

corona10 committed:
Minor refactoring

corona10 commented on Modules/posixmodule.c:
```diff
@@ -14335,6 +14335,22 @@ os_get_terminal_size_impl(PyObject *module, int fd)
+[clinic start generated code]*/
+
+static PyObject *
+os__get_cpu_count_config_impl(PyObject *module)
```
This function will return "process" for the process mode with the separate PR.

corona10 replied:
Which will be, but will be done with a separate PR.

```python
if _exists('sched_getaffinity') and _get_cpu_count_config() != 'overridden':
    def process_cpu_count():
        """
        Get the number of CPUs of the current process.
        Return the number of logical CPUs usable by the calling thread of the
        current process. Return None if indeterminable.
        """
        return len(sched_getaffinity(0))

   if _get_cpu_count_config() == "process":
       cpu_count = process_cpu_count
```

corona10 committed:
Revert to PYTHONCPUCOUNT

corona10 committed:
fix

corona10 committed:
Use PYTHON_CPU_COUNT

gpshead commented on Lib/os.py:
```diff
@@ -55,6 +55,7 @@ def _get_exports_list(module):
     from posix import *
     try:
         from posix import _exit
+        from posix import _get_cpu_count_config
```
This means it is only possible to override cpu_count on posix?

corona10 replied:
Reasonable!

gpshead commented on Modules/posixmodule.c:
```diff
@@ -14335,6 +14335,22 @@ os_get_terminal_size_impl(PyObject *module, int fd)
+/*[clinic input]
+os._get_cpu_count_config
+
+Private function to get PyConfig.cpu_count is overridden or not
```
This doesn't belong in posixmodule, it is not platform specific.  `Python/sysmodule.c` perhaps?

and rather than an internal API returning two magic strings, how about an internal API that just returns the overridden value. Code calling it should only use it if it returns an int > 0, anything <= 0 means invalid/default/not-overridden.

That way os_cpu_count_impl below could just call the c impl function from sysmodule directly rather than having a copy of similar code.

corona10 replied:
> and rather than an internal API returning two magic strings, how about an internal API that just returns the overridden value. Code calling it should only use it if it returns an int > 0, anything <= 0 means invalid/default/not-overridden.

Well, Victor and I worry about someone get the CPU count from the `_get_cpu_count_config()`...
I want to prevent the API from being used for root of data :(

corona10 replied:
FYI, -2 will be the process mode.

corona10 replied:
I am worrying about this kinds of code, we never expect but somebody can do.

```python
cpu_count = sys._get_cpu_count_config() if sys._get_cpu_count_config() > 0 else os.process_cpu_count()
```

gpshead replied:
They could also `import posix` and do that from this location. I'm not worried about anyone using an `_` prefixed private undocumented api to shoot themselves in the foot.

Bikeshedding could come up with other more or less tempting names, but I don't think it matters too much. If it starts with `_` and we keep it undocumented, it isn't for public consumption.  `_get_overridden_cpu_count()` `_get_eaten_by_a_grue()` `_get_do_not_press_this_button_press_count()` ...

we've got plenty of other _ functions in this category in modules already.

corona10 replied:
>  I'm not worried about anyone using an _ prefixed private undocumented api to shoot themselves in the foot.
> we've got plenty of other _ functions in this category in modules already.

Okay!

corona10 committed:
Address Greg's review

corona10 commented on Python/sysmodule.c:
```diff
@@ -2290,6 +2290,23 @@ sys__getframemodulename_impl(PyObject *module, int depth)
+[clinic start generated code]*/
+
+static PyObject *
+sys__get_cpu_count_config_impl(PyObject *module)
```
@gpshead 

I wrote the comment about the need of magic string in https://github.com/python/cpython/pull/109667#discussion_r1343351179

PTAL

corona10 replied:
Agreement with https://github.com/python/cpython/pull/109667#discussion_r1343372035

corona10 committed:
Address Greg's code review

corona10 committed:
nit

gpshead committed:
Include multiprocessing in docs, reword.

Adds PYTHON_CPU_COUNT help text.
Mentions multiprocessing.cpu_count.
rewords a few statements.

corona10 committed:
Merge remote-tracking branch 'upstream/main' into gh-109595

corona10 committed:
Merge remote-tracking branch 'upstream/main' into gh-109595

vstinner reviewed (approved):
LGTM. I see that you addressed the two @gpshead's comments. You can merge.
